### ISSUE TITLE
Internal: stack overwrite on _call_native_mb_ret with godot::Error return value  

### ISSUE THREAD

**Issue by derammo** [godotengine/godot#64455]
### Godot version

custom

### System information

Windows 10 Visual Studio 2019

### Issue description

for `R` is `godot::Error`, a 4-byte enum:

```
template <class R, class... Args>
R _call_native_mb_ret(const GDNativeMethodBindPtr mb, void *instance, const Args &...args) {
	R ret;
	std::array<const GDNativeTypePtr, sizeof...(Args)> mb_args = { { (const GDNativeTypePtr)args... } };
	internal::gdn_interface->object_method_bind_ptrcall(mb, instance, mb_args.data(), &ret);
	return ret;
}
```

before object_method_bind_ptrcall:

![image](https://user-images.githubusercontent.com/817160/184677799-3987eba0-7425-44de-bf89-9d849a38d7ec.png)

after object_method_bind_ptrcall:

![image](https://user-images.githubusercontent.com/817160/184677840-88c67ff7-1a98-47fc-9630-c1ddfa6f9bc0.png)

It has written 8 bytes and triggered stack guard detection in Visual Studio.   I believe this is real and it may be caused by VARIANT_ENUM_CAST assuming all `enum` are `int64_t` and thus overwriting the 4-byte stack variable on `encode`.  reduz would probably know instantly if this is a real issue.

If it becomes relevant, I can try to reproduce this with an official build, but it is hard to do right now because of problems in GDExtension that require me to run various patches.   In any case this would be confirmed and fixed by inspection if it is previously unknown.  

### Steps to reproduce

Debugging extensions code, and by inspection.

**Comment by derammo** [comment]
I put this here for tracking if this is not a known issue.  I can't really provide a good repro for it, which would also be a waste of time.  Either this is a code bug, or it is not, and @reduz would know the difference pretty much instantly.

**Comment by RiederAlex** [comment]
I've also hit the issue while playing around with godot-cpp for 4.0.
For godot-cpp: I've fixed it locally by explicitly specifying the underlying enum types for the generated enumerations (global_constants.hpp), so they are all of type int64_t `enum Error` -> `enum Error : int64_t`, which could also be one potential solution for the enums in this repository.

**Comment by derammo** [comment]
Yes, that would be a work-around.  Could it be combined with a static assertion to make sure it isn't being used that way anywhere else?

**Comment by derammo** [comment]
By the way, I strongly suggest this should be fixed by correcting the encoding.  I am just trying to confirm that based on what I saw in the code, the work-around would work.

**Comment by derammo** [comment]
Code that forces the enums to be assumed wide is here:
https://github.com/godotengine/godot/commit/95088f6bfaad8af476198f5e7d52baf06b86edf1

Unfortunately, this means that the encoding blindly writes `int64_t` into the `void*`, even if the underlying type is a narrower `enum`.

**Comment by derammo** [comment]
https://github.com/godotengine/godot/pull/50931

**Comment by RiederAlex** [comment]
If this issue causes the enums to have explicit types I would also propose to think if int64_t is really needed in this case and if a smaller data-type would suffice - as an enum with this much entries does not really make sense (except maybe for really huge bitflags?) plus encoding/decoding/saving/loading/network all use double the amount of data needed.

**Comment by DNKpp** [comment]
> If this issue causes the enums to have explicit types I would also propose to think if int64_t is really needed in this case and if a smaller data-type would suffice - as an enum with this much entries does not really make sense (except maybe for really huge bitflags?) plus encoding/decoding/saving/loading/network all use double the amount of data needed.

In my opinion such assumptions are always the wrong direction. Its totally fine assigning a ``2^20`` to an enum type, as long as it has enough bits, even if that enum (class) type has no named constant values. I don't see why those enum types shouldn't use the corresponding integer with equal bit size.

**Comment by DNKpp** [comment]
Even if its not pretty, but this should already be sufficent.
```cpp
template <class TEnum>
using int_rep_t = std::conditional_t<
    sizeof(TEnum) == 1,
    std::uint8_t,
    std::conditional_t<
        sizeof(TEnum) == 2,
        std::uint16_t,
        std::conditional_t<
            sizeof(TEnum) == 4,
            std::uint32_t,
            std::uint64_t
            >
        >
    >;
```

**Comment by derammo** [comment]
Damn it, I was looking forward to messing with conditional template stuff but you are already done.   Are you planning to PR this?

**Comment by DNKpp** [comment]
> Damn it, I was looking forward to messing with conditional template stuff but you are already done. Are you planning to PR this?

Currently not, no. As I do not have much insight into the godot codebase, I would have to do some code digging in forehand.

**Comment by derammo** [comment]
Ok I might steal your code then if I get motivated.  I have the advantage of being set up to test already.   Unless @RiederAlex wants to do it?

**Comment by RiederAlex** [comment]
Explicit underlying enum specification would probably still be needed (as otherwise different compilers could choose different types, causing problems) + I guess if its not always the same size this information would need to be included in the "extension_api.json".

**Comment by DNKpp** [comment]
> Explicit underlying enum specification would probably still be needed (as otherwise different compilers could choose different types, causing problems) + I guess if its not always the same size this information would need to be included in the "extension_api.json".

That's the reason why I said, I would have to do some code digging before I could provide a PR. From a quick look into ``extension_api.json`` it seems, that many types have a explicit ``size`` property. Couldn't that be adapted for core enum types aswell? Is that ``extionsion_api.json`` a generated file or is it manually modified?

**Comment by derammo** [comment]
@RiederAlex Wow, great point.

@DNKpp It is generated.  The code I linked and that we are talking about modifying is the glue code that marshals information across the extension boundary.  So yes, I believe you are correct, and we need that info in extension_api because it is ABI (not just API.)  I guess this means either all enums for Extension must be widened to 64 bit or must have a declared width in the API JSON.

**Comment by derammo** [comment]
Please also note that the original code as initially developed just had `int`.  The widening to `int64_t` was done later by someone else for an unrelated reason.  So there definitely was no conscious decision to just assume all `enum` are wide.    Considering it is so much easier and safer to do, do you think it is best to start with just ensuring via static assert that all `enum` in the ABI are wide?   Then we'd find out very quickly how many types are affected :) and it would prevent this from ever happening again.   If there is really only the one (`godot::Error`) or a few then we could easily specialize the templates for that type and allow it to remain 4 bytes.

**Comment by DNKpp** [comment]
Note: didn't knew about [this](https://en.cppreference.com/w/cpp/types/underlying_type). I think that should be prefered over my solution above.

**Comment by derammo** [comment]
Good find but we still need to control what size it is explicitly somehow because different compilers can be used on opposite sides of the extension boundary.  We can’t assume they will choose the same enum size?

**Comment by DNKpp** [comment]
@derammo  Yeah, that's why I prefixed my message with "note:". Do you have a hint where to find the generating code of ``extionsion_api.json``?
